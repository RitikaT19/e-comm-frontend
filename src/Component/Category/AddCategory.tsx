import React, { useState } from "react";
import Button from "../common/Button/Button";
import { Textfield } from "../common/Textfield/Textfield";
import "../styles/add-category.css";
import cross from "../../assets/Icons/cross.png";
import { Loading } from "../common/Loading/Loading";
import { isCategoryValid } from "../../helpers/validate";

// Define all the props required by the component
interface Props {
  handleAddCategory: any;
  CrossIconClick: any;
  isEdit: boolean;
  showLoader: boolean;
  categoryDetail: any;
  handleEditCategory: any;
  errorMessage: any;
  successMessage: any;
}

export const AddCategory: React.FC<Props> = ({
  handleAddCategory,
  CrossIconClick,
  isEdit,
  showLoader,
  categoryDetail,
  handleEditCategory,
  errorMessage,
  successMessage,
}) => {
  // stores name
  const [name, setName] = useState<string>(isEdit ? categoryDetail.name : "");
  // stores name error
  const [nameError, setNameError] = useState<string>("");
  // stores empty field error
  const [showEmptyFieldError, setShowEmptyFieldError] = useState(false);

  const checkCategory = async () => {
    const validateCategoryResult = isCategoryValid(name);
    if (!validateCategoryResult) {
      setNameError("Please enter valid category");
    } else {
      setNameError("");
    }
  };
  console.log(successMessage, "error from category");

  // function for when submitCategory button is clicked
  const submitCategory = async () => {
    if (!name) {
      setShowEmptyFieldError(true);
    } else {
      // call handle add category with name
      await handleAddCategory({ name });
      console.log(name, "name from add category");
    }
  };
  const editCategoryButtonClick = async () => {
    const data = { id: categoryDetail?._id, name };
    await handleEditCategory({ id: categoryDetail?._id, name });
    console.log(categoryDetail, "categoty etai");
  };
  return (
    <div>
      {/* if show loader is true, show loader */}
      {showLoader ? (
        <div className="add-category-loader">
          <Loading />
        </div>
      ) : (
        <div className="add-category-container">
          <div className="add-category-background">
            <div className="add-category-div">
              <div className="close-button">
                {/* cross button */}
                <img src={cross} alt="cross" onClick={CrossIconClick} />
              </div>
              {/* input for category name */}
              <Textfield
                id="category-name"
                label={isEdit ? "Edit Category" : "Add Category"}
                placeholder="Enter category name here"
                value={name}
                onChange={(e: any) => setName(e.target.value)}
                error={nameError}
                onBlur={checkCategory}
              />
            </div>
            {errorMessage ? (
              <p className="add-user-error"> {errorMessage}</p>
            ) : successMessage ? (
              <p className="add-user-success">{successMessage}</p>
            ) : (
              showEmptyFieldError && (
                <p className="add-user-error">Please fill all the fields</p>
              )
            )}
            {/* button for submitting category */}
            <Button
              id={isEdit ? "edit-category-button" : "add-category-button"}
              value={isEdit ? "Edit Category" : "Add Category"}
              handleClick={isEdit ? editCategoryButtonClick : submitCategory}
            />
          </div>
        </div>
      )}
    </div>
  );
};
