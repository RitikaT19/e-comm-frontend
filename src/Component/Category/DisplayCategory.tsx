import React from "react";
import Button from "../common/Button/Button";

interface Props {
    categoryDetails : any,
    addCategoryClick: any

}
export const DisplayCategory: React.FC<Props> = ({addCategoryClick, categoryDetails}) =>{
    return(
        <div className = "display-category-container">
            <div className = "display-category-action">
            <p className = "display-category-header">
                Categories
            </p>
            <Button
            id = "display-category-button"
            value = "Add Category"
            handleClick = {addCategoryClick}/>
            </div>
            
         <div className = "display-category">
            {categoryDetails?.map((item: any, index: number)=>(
                <p id = {"display-category-name"}>{item.name}</p>
            ))}
        </div> 
        </div>
    )
}