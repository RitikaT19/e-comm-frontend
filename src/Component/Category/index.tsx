import React, { useContext, useEffect, useState } from "react";
import { SideBar } from "../common/SideBar/Sidebar";
import {
  createNewCategory,
  getCategory,
  deleteCategory,
  editCategory,
} from "../../actions/category";
import { CategoryContext } from "../../contexts/Category";
import { LoadingContext } from "../../contexts/Loading";
import "../styles/category.css";
import { DisplayCategory } from "./DisplayCategory";
import { AddCategory } from "./AddCategory";
import Button from "../common/Button/Button";

export const Category: React.FC = () => {
  //rename state and dispatch as categoryState and categoryDispatch respectively
  const { state: categoryState, dispatch: categoryDispatch } =
    useContext(CategoryContext);
  const { dispatch: loadingDispatch } = useContext(LoadingContext);
  // stores flag for showing add category page
  const [showAddCategoryPage, setShowAddCategoryPage] = useState(false);
  // flag for display category loader
  const [showDisplayCategoryLoader, setDisplayCategoryShowLoader] = useState(false)
  // flag for display category loader
  const [showAddCategoryLoader, setAddCategoryShowLoader] = useState(false)
  // stores the state for edit category flag
  const [editCategoryData, setEditCategoryData] = useState<boolean>(false)
  // stored the detail of a single category
  const [categoryData, setCategoryData] = useState<any>()
  
  
  // toggle function for showing add category page
  const addCategoryToggle = () => {
    // toggles add category page
    setShowAddCategoryPage(!showAddCategoryPage);
    // sets edit flag as false
    setEditCategoryData(false)
    // sets category data as empty
    setCategoryData({});
  };

  // function for fetching category
  const fetchCategory = async () => {
    setDisplayCategoryShowLoader(true)
    // call getCategory action
    await getCategory()(categoryDispatch, loadingDispatch).then(()=>{
      setDisplayCategoryShowLoader(false)
    });
  };

  // call useEffect when fetchCategory function is successful
  useEffect(() => {
    fetchCategory();
  }, []);

  // function for adding category
  const addCategory = async (data: string) => {
    setAddCategoryShowLoader(true)
    // call createNewCategory action
    await createNewCategory(data)(categoryDispatch, loadingDispatch).then(()=>{
      setAddCategoryShowLoader(false)
    });
    // call fetchCategory everytime after a new category is added
    fetchCategory();
    setShowAddCategoryPage(false)
  };

  // function for deleting a category
  const removeCategory = async (id: any) => {
    // call deleteCategory action
    await deleteCategory(id)(categoryDispatch, loadingDispatch);
    //  call fetchCategory after deleting a category
    fetchCategory();
  };

  // function for updating category
  const updateCategory = async(data: any) =>{
    setAddCategoryShowLoader(true)
    // call action editCategory
    await editCategory(data)(categoryDispatch, loadingDispatch).then(()=>(
    setAddCategoryShowLoader(false)
    ))
    fetchCategory()
    setShowAddCategoryPage(false)
  };

  const updateCategoryIconClick = (data: any) =>{
    // sets the category detail for the card clicked
    setCategoryData(data);
    // sets the setShowAddCategoryPage true 
    setShowAddCategoryPage(true)
    // sets the edit flag true
    setEditCategoryData(true)
    
  }
  return (
    <div className="category-main-div">
      <SideBar />
      <div className="display-category-action">
        <p className="display-category-header">
          <b> Categories</b>
        </p>
        <Button
          id="display-category-button"
          value="Add Category"
          handleClick={addCategoryToggle}
        />
      </div>

      {console.log(categoryState.error, "...")}

      <div className="category-container-div">
        {/* if showAddCategoryPage is true then render AddCategory page else render display category page */}
        {showAddCategoryPage ? (
          <AddCategory
            CrossIconClick={addCategoryToggle}
            handleAddCategory={addCategory}
            isEdit = {editCategoryData}
            showLoader = {showAddCategoryLoader}
            categoryDetail = {categoryData}
            handleEditCategory = {updateCategory}
            errorMessage = {editCategoryData ? categoryState.updateCategoryError : categoryState.error}
            successMessage = {editCategoryData ? categoryState.updateCategorySuccess : categoryState.createCategorySuccess}
          />
        ) : (
          <DisplayCategory
            categoryInfo={categoryState.fetchCategorySuccess}
            deleteCategory={removeCategory}
            updateCategoryIconClick={updateCategoryIconClick}
            showLoader = {showDisplayCategoryLoader}
          />
        )}
      </div>
    </div>
  );
};
