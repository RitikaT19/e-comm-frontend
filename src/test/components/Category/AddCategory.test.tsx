import React, { Component } from "react";
import { mount, ReactWrapper } from "enzyme";
import { AddCategory } from "../../../component/Category/AddCategory";

describe("testing Add Category", () => {
  let wrapper: ReactWrapper<any, Readonly<{}>, Component<{}, {}, any>>;
  let wrapper2: ReactWrapper<any, Readonly<{}>, Component<{}, {}, any>>;

  const handleAddCategory: any = jest.fn();
  const CrossIconClick: any = jest.fn();
  const isEdit: boolean = false;
  const showLoader: boolean = false;
  const handleEditCategory: any = jest.fn();
  const errorMessage = "";
  const successMessage = "";
  const categoryDetail: any = {
    name: "Test",
  };
  const categoryDetail1: any = {
    name: "Test",
  };
  const defaultProps = {
    handleAddCategory,
    CrossIconClick,
    isEdit,
    showLoader,
    categoryDetail,
    handleEditCategory,
    errorMessage,
    successMessage,
  };

  const defaultProps1 = {
    handleAddCategory,
    CrossIconClick,
    isEdit: true,
    showLoader,
    categoryDetail: categoryDetail1,
    handleEditCategory,
    errorMessage,
    successMessage,
  };

  beforeEach(() => {
    wrapper = mount(<AddCategory {...defaultProps} />);
  });
  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot;
  });

  it("should have 1 button component", () => {
    expect(wrapper.find("Button")).toHaveLength(1);
  });
  it("should have 1 TextField component", () => {
    let textFields = wrapper.find("TextField").length;
    expect(wrapper.find("TextField")).toHaveLength(textFields);
  });
  it("should call the handle change prop when a user types something in the password input field", () => {
    wrapper.find("#add-category-button_element").simulate("click");
    const event = { target: { value: "tests" } };
    const event2 = { target: { value: "test 123" } };
    wrapper.find("#category-name_element").simulate("change", event2);
    wrapper.find("#category-name_element").simulate("change", event);
  });

  it("should call the handle change prop when a user types something in the password input field", () => {
    wrapper2 = mount(<AddCategory {...defaultProps1} />);
    wrapper.find("#add-category-button_element").simulate("click");
    const event = { target: { value: "tests" } };
    const event2 = { target: { value: "test 123" } };
    wrapper.find("#category-name_element").simulate("change", event2);
    wrapper.find("#category-name_element").simulate("change", event);
  });

  it("should check for Loader visible", () => {
    wrapper.setProps({ showLoader: true });
    expect(wrapper.find(".add-category-loader")).toHaveLength(1);
  });

});
