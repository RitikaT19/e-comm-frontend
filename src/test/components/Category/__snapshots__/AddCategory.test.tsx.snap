// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Add Category should match snapshot when edit flag is true 1`] = `
<AddCategory
  CrossIconClick={[MockFunction]}
  categoryDetail={
    Object {
      "name": "Test",
    }
  }
  errorMessage=""
  handleAddCategory={[MockFunction]}
  handleEditCategory={[MockFunction]}
  isEdit={true}
  showLoader={false}
  successMessage=""
>
  <div>
    <div
      className="add-category-container"
    >
      <div
        className="add-category-div"
      >
        <div
          className="close-button"
        >
          <img
            alt="cross"
            onClick={[MockFunction]}
            src="cross.png"
          />
        </div>
        <Textfield
          error=""
          id="category-name"
          label="Edit Category"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Enter category name here"
          value="Test"
        >
          <div
            className="textfield-container"
          >
            <label
              htmlFor="input-field"
            >
              Edit Category
            </label>
            <input
              className="textfield"
              id="category-name_element"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="Enter category name here"
              value="Test"
            />
          </div>
        </Textfield>
      </div>
      <Button
        handleClick={[Function]}
        id="edit-category-button"
        value="Edit Category"
      >
        <button
          className="button"
          id="edit-category-button_element"
          onClick={[Function]}
        >
          Edit Category
        </button>
      </Button>
    </div>
  </div>
</AddCategory>
`;
